### PR TITLE
feat(ciapp): detect code owners for PyTest tests

### PR DESCRIPTION
**Pull request by jirikuncar**
Read CODEOWNERS in `pytest` integration. If it's present, match the test file path and add its owners as a `test.codeowners` span tag.

## Motivation

Allow owners of failed tests to be notified.

## Note

It's been tested with `datadog-api-client-python` in https://github.com/DataDog/datadog-api-client-python/pull/971.

## Checklist
- [x] Added to the correct milestone.
- [x] Tests provided or description of manual testing performed is included in the code or PR.
- [ ] ~Library documentation is updated.~
- [ ] ~[Corp site](https://github.com/DataDog/documentation/) documentation is updated (link to the PR).~

### CODE REVIEW

File: tests/contrib/pytest/test_pytest.py [hunk 1]
```diff
@@ -698,6 +698,36 @@ def test_will_work():
 
         assert test_span.get_tag(test.FRAMEWORK_VERSION) == pytest.__version__
 
+    def test_pytest_will_report_codeowners(self):
```
Comment (brettlangdon) [inline tests/contrib/pytest/test_pytest.py:701]: can we add tests with missing or malformed CODEOWNERS, or with files that do not match any code owners?

File: ddtrace/internal/codeowners.py [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+        self.patterns = []  # type: List[Tuple[re.Pattern, List[str]]]
+        self.parse()
+
+    def location(self, cwd=None):
+        # type: (Optional[str]) -> Optional[str]
+        """Return the location of the CODEOWNERS file."""
+        cwd = cwd or os.getcwd()
+        for location in self.KNOWN_LOCATIONS:
+            path = os.path.join(cwd, location)
+            if os.path.exists(path):
```
Comment (brettlangdon) [inline ddtrace/internal/codeowners.py:143]:
```suggestion
            if os.path.isfile(path):
```

? 🤷🏻 either is probably fine. this suggestion is not required, feel free to dismiss

File: ddtrace/internal/codeowners.py [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+                return path
+        raise ValueError("CODEOWNERS file not found")
+
+    def parse(self):
+        # type: () -> None
+        """Parse CODEOWNERS file and store the lines and regexes."""
+        with open_text(self.path) as f:
```
Comment (brettlangdon) [inline ddtrace/internal/codeowners.py:150]: why this vs just `open(self.path)` ?

Comment (brettlangdon) [reply ddtrace/internal/codeowners.py:150]: or at least a comment on why we should use `open_text` would be good

File: ddtrace/internal/codeowners.py [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+            for line in f.readlines():
+                line = line.strip()
+                if line == "":
+                    continue
+                # Lines starting with '#' are comments.
+                if line.startswith("#"):
+                    continue
+                if line.startswith("[") and line.endswith("]"):
+                    continue
+                if line.startswith("^[") and line.endswith("]"):
```
Comment (brettlangdon) [inline ddtrace/internal/codeowners.py:161]: what do these mean?

File: ddtrace/internal/codeowners.py [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+                    continue
+
+                elements = iter(line.split())
+                path = next(elements, None)
+                if path is None:
+                    continue
+
+                owners = []
```
Comment (brettlangdon) [inline ddtrace/internal/codeowners.py:169]: set? maybe a tiny optimization for a rare case, but wdyt?

File: ddtrace/internal/codeowners.py [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+                for owner in elements:
+                    if owner:
+                        owners.append(owner)
+
+                if not owners:
+                    continue
+                patterns.append((path_to_regex(path), owners))
+            # Order is important. The last matching pattern has the most precedence.
```
Comment (brettlangdon) [inline ddtrace/internal/codeowners.py:177]: does this order matter for ciapp as well?

**Review by brettlangdon — COMMENTED** [review]
tiny non-blocking nits, but I would like to see more tests for ensuring we handle CODEOWNERS files under different scenarios

File: ddtrace/internal/codeowners.py (hunk 5 above)
Comment (jirikuncar) [reply ddtrace/internal/codeowners.py:169]: I would prefer to send exactly what is in the CODEOWNERS file. It is a very very rare to have duplicated values and it does not compensate building a set.

File: ddtrace/internal/codeowners.py (hunk 4 above)
Comment (jirikuncar) [reply ddtrace/internal/codeowners.py:161]: I can add links to the documentation:
```suggestion
                if line.startswith("[") and line.endswith("]"):
                    # https://docs.gitlab.com/ee/user/project/code_owners.html#organize-code-owners-by-putting-them-into-sections
                    continue
                if line.startswith("^[") and line.endswith("]"):
                    # https://docs.gitlab.com/ee/user/project/code_owners.html#make-a-code-owners-section-optional
```
or simple comments
```suggestion
                if line.startswith("[") and line.endswith("]"):
                    # found a code owners section
                    continue
                if line.startswith("^[") and line.endswith("]"):
                    # found an optional code owners section
```

File: tests/contrib/pytest/test_pytest.py (hunk 1 above)
Comment (jirikuncar) [reply tests/contrib/pytest/test_pytest.py:701]: I have added a test and `try/catch` for line parsing 👍

**Review by majorgreys — APPROVED** [review]

File: ddtrace/internal/codeowners.py [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+                    continue
+
+                elements = iter(line.split())
+                path = next(elements, None)
+                if path is None:
+                    continue
+
+                try:
+                    pattern = path_to_regex(path)
+                except ValueError:
```
Comment (P403n1x87) [inline ddtrace/internal/codeowners.py:171]:
```suggestion
                try:
                    pattern = path_to_regex(line.split()[0])
                except ValueError, IndexError:
```
?

File: ddtrace/internal/codeowners.py [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+                    continue
+
+                owners = []
+                for owner in elements:
+                    if owner:
+                        owners.append(owner)
```
Comment (P403n1x87) [inline ddtrace/internal/codeowners.py:177]:
```suggestion
                owners = [owner for owner in elements if owner]
```

File: ddtrace/internal/codeowners.py [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+            self.patterns = patterns
+
+    def of(self, path):
+        # type: (str) -> List[str]
+        """Return code owners for a given path.
+
+        :param path: path to check
+        :return: list of file code owners identified by the given path
+        """
+        for pattern, owners in self.patterns:
```
Comment (P403n1x87) [inline ddtrace/internal/codeowners.py:185]: is it worth caching the result of this method? do we expect these lookups to be repeated frequently for the same path?

File: releasenotes/notes/feature-ciapp-detect-code-owners-for-PyTest-tests-f8f66db2dc02ae91.yaml [hunk 10]
```diff
@@ -0,0 +1,4 @@
+---
+features:
+  - |
+    ciapp: detect code owners of PyTest tests
```
Comment (P403n1x87) [inline releasenotes/notes/feature-ciapp-detect-code-owners-for-PyTest-tests-f8f66db2dc02ae91.yaml:4]: missing newline

File: ddtrace/internal/codeowners.py (hunk 9 above)
Comment (jirikuncar) [reply ddtrace/internal/codeowners.py:185]: I don't think it is necessary at the moment. Each integration can have its caching strategy if needed.

File: releasenotes/notes/feature-ciapp-detect-code-owners-for-PyTest-tests-f8f66db2dc02ae91.yaml (hunk 10 above)
Comment (jirikuncar) [reply releasenotes/notes/feature-ciapp-detect-code-owners-for-PyTest-tests-f8f66db2dc02ae91.yaml:4]:
```suggestion
    ciapp: detect code owners of PyTest tests

```

File: tests/contrib/pytest/test_pytest.py [hunk 11]
```diff
@@ -698,6 +698,36 @@ def test_will_work():
...
+        file_names.append(os.path.basename(py_team_b_file.strpath))
+        codeowners = "* @default-team\n{0} @team-b @backup-b".format(os.path.basename(py_team_b_file.strpath))
+        self.testdir.makefile("", CODEOWNERS=codeowners)
+
+        self.inline_run("--ddtrace", *file_names)
+        spans = self.pop_spans()
+
+        assert len(spans) == 2
+        assert json.loads(spans[0].get_tag(test.CODE_OWNERS)) == ["@default-team"], spans[0]
+        assert json.loads(spans[1].get_tag(test.CODE_OWNERS)) == ["@team-b", "@backup-b"], spans[1]
```
Comment (jirikuncar) [inline tests/contrib/pytest/test_pytest.py:729]:
```suggestion
        assert json.loads(spans[0].get_tag(test.CODEOWNERS)) == ["@default-team"], spans[0]
        assert json.loads(spans[1].get_tag(test.CODEOWNERS)) == ["@team-b", "@backup-b"], spans[1]
```

**Review by majorgreys — APPROVED** [review]

**Review by Kyle-Verhoog — APPROVED** [review]

**Pull request merged**